commit 45af3fb4a018ef84bf1c9f2dfbd887a41242e77f
Author: Glen Turner <gdt@gdt.id.au>
Date:   Wed Feb 27 04:32:36 2013 +0000

    usb/net/asix_devices: Add USBNET HG20F9 ethernet dongle
    
    This USB ethernet adapter was purchased in anodyne packaging
    from the computer store adjacent to linux.conf.au 2013 in
    Canberra (Australia). A web search shows other recent
    purchasers in Lancaster (UK) and Seattle (USA). Just like an
    emergent virus, our age of e-commerce and airmail allows
    underdocumented hardware to spread around the world instantly
    using the vector of ridiculously low prices.
    
    Paige Thompson, infected via eBay, discovered that the HG20F9
    is a copy of the Asix 88772B; many viruses copy the RNA of
    other viruses. See Paige's work at
    <https://github.com/paigeadele/HG20F9>.
    This patch uses her discovery to update the restructured Asix
    driver in the current kernel.
    
    Just as some viruses inhabit seemingly-healthy cells, the
    HG20F9 uses the Vendor ID 0x066b assigned to Linksys Inc.
    For the present there is no clash of Product ID 0x20f9.
    
    Signed-off-by: Glen Turner <gdt@gdt.id.au>
    Signed-off-by: David S. Miller <davem@davemloft.net>

diff --git a/drivers/net/usb/asix_devices.c b/drivers/net/usb/asix_devices.c
index 2205dbc..7097534 100644
--- a/drivers/net/usb/asix_devices.c
+++ b/drivers/net/usb/asix_devices.c
@@ -924,6 +924,29 @@ static const struct driver_info ax88178_info = {
 	.tx_fixup = asix_tx_fixup,
 };
 
+/*
+ * USBLINK 20F9 "USB 2.0 LAN" USB ethernet adapter, typically found in
+ * no-name packaging.
+ * USB device strings are:
+ *   1: Manufacturer: USBLINK
+ *   2: Product: HG20F9 USB2.0
+ *   3: Serial: 000003
+ * Appears to be compatible with Asix 88772B.
+ */
+static const struct driver_info hg20f9_info = {
+	.description = "HG20F9 USB 2.0 Ethernet",
+	.bind = ax88772_bind,
+	.unbind = ax88772_unbind,
+	.status = asix_status,
+	.link_reset = ax88772_link_reset,
+	.reset = ax88772_reset,
+	.flags = FLAG_ETHER | FLAG_FRAMING_AX | FLAG_LINK_INTR |
+	         FLAG_MULTI_PACKET,
+	.rx_fixup = asix_rx_fixup_common,
+	.tx_fixup = asix_tx_fixup,
+	.data = FLAG_EEPROM_MAC,
+};
+
 extern const struct driver_info ax88172a_info;
 
 static const struct usb_device_id	products [] = {
@@ -1063,6 +1086,14 @@ static const struct usb_device_id	products [] = {
 	/* ASIX 88172a demo board */
 	USB_DEVICE(0x0b95, 0x172a),
 	.driver_info = (unsigned long) &ax88172a_info,
+}, {
+	/*
+	 * USBLINK HG20F9 "USB 2.0 LAN"
+	 * Appears to have gazumped Linksys's manufacturer ID but
+	 * doesn't (yet) conflict with any known Linksys product.
+	 */
+	USB_DEVICE(0x066b, 0x20f9),
+	.driver_info = (unsigned long) &hg20f9_info,
 },
 	{ },		// END
 };
